File: src/main/java/org/genyris/interp/builtin/TagFunction.java
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.interp.builtin;

import org.genyris.core.DynamicSymbol;
import org.genyris.core.Exp;
import org.genyris.core.SimpleSymbol;
import org.genyris.core.StandardClass;
import org.genyris.exception.GenyrisException;
import org.genyris.interp.ApplicableFunction;
import org.genyris.interp.Closure;
import org.genyris.interp.Environment;
import org.genyris.interp.GenyrisTypeMismatchException;
import org.genyris.interp.Interpreter;
import org.genyris.interp.UnboundException;

public class TagFunction extends ApplicableFunction {

    public TagFunction(Interpreter interp) {
        super(interp, "tag", true);
    }

    private SimpleSymbol VALIDATE() {
        return _interp.getSymbolTable().VALIDATE();
    }

    public Exp bindAndExecute(Closure proc, Exp[] arguments,
            Environment environment) throws GenyrisException {
        checkArguments(arguments, 2);
        Class[] types = { StandardClass.class };
        checkArgumentTypes(types, arguments);
        Exp object = arguments[1];
        StandardClass klass = (StandardClass) arguments[0];
        callValidator(environment, object, klass);
        object.addClass(klass);
        return object;
    }

    // TODO move these into ClassWrapper:
    public void callValidator(Environment environment, Exp object,
            StandardClass klassobject) throws GenyrisException {
        // TODO DRY
        Exp NIL = environment.getNil();
        Exp validator = null;
        try {
            validator = klassobject.lookupVariableValue(new DynamicSymbol(VALIDATE()));
        } catch (UnboundException ignore) {
            return;
        }
        Exp args[] = new Exp[1];
        args[0] = object;
        Exp result = validator.applyFunction(klassobject, args);
        if (result == NIL) {
            throw new GenyrisException("class " + klassobject.getClassName()
                    + " validator error for object " + object);
        }
    }

    public static void validateObjectInClass(Environment environment,
            Exp object, StandardClass klassobject) throws GenyrisException {
        Exp NIL = environment.getNil();
        Exp validator = null;
        try {
            validator = klassobject.lookupVariableValue(new DynamicSymbol(VALIDATE(environment)));
        } catch (UnboundException ignore) {
            if (!klassobject.isInstance(object)) {
              throw new GenyrisTypeMismatchException("validator error: object "  + object +
                        " is not tagged with " + klassobject.getClassName() );
            }
        }
        if (validator != null) {
            Exp args[] = new Exp[1];
            args[0] = object;
            Exp result = validator.applyFunction(klassobject, args);
            if (result == NIL) {
                throw new GenyrisException("class " + klassobject.getClassName()
                        + " validator error for object " + object);
            }
        }
    }

    private static SimpleSymbol VALIDATE(Environment environment) {
        return environment.getSymbolTable().VALIDATE();
    }

}

File: src/org/genyris/interp/builtin/TagFunction.kt
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.interp.builtin

import org.genyris.core.DynamicSymbol
import org.genyris.core.Exp
import org.genyris.core.SimpleSymbol
import org.genyris.core.StandardClass
import org.genyris.exception.GenyrisException
import org.genyris.interp.*

class TagFunction(interp: Interpreter) : ApplicableFunction(interp, "tag", true) {
    private fun VALIDATE(): SimpleSymbol? {
        return _interp.getSymbolTable().VALIDATE()
    }

    @kotlin.Throws(GenyrisException::class)
    override fun bindAndExecute(
        proc: Closure?, arguments: Array<Exp>,
        environment: Environment
    ): Exp {
        checkArguments(arguments, 2)
        val types = arrayOf<Class<*>?>(StandardClass::class.java)
        checkArgumentTypes(types, arguments)
        val `object` = arguments[1]
        val klass = arguments[0] as StandardClass
        callValidator(environment, `object`, klass)
        `object`.addClass(klass)
        return `object`
    }

    // TODO move these into ClassWrapper:
    @kotlin.Throws(GenyrisException::class)
    fun callValidator(
        environment: Environment, `object`: Exp?,
        klassobject: StandardClass
    ) {
        // TODO DRY
        val NIL: Exp? = environment.getNil()
        var validator: Exp? = null
        try {
            validator = klassobject.lookupVariableValue(DynamicSymbol(VALIDATE()))
        } catch (ignore: UnboundException) {
            return
        }
        val args: Array<Exp?>? = arrayOfNulls<Exp>(1)
        args!![0] = `object`
        val result = validator.applyFunction(klassobject, args)
        if (result === NIL) {
            throw GenyrisException(
                ("class " + klassobject.getClassName()
                        + " validator error for object " + `object`)
            )
        }
    }

    companion object {
        @kotlin.Throws(GenyrisException::class)
        fun validateObjectInClass(
            environment: Environment,
            `object`: Exp, klassobject: StandardClass
        ) {
            val NIL: Exp? = environment.getNil()
            var validator: Exp? = null
            try {
                validator = klassobject.lookupVariableValue(DynamicSymbol(VALIDATE(environment)))
            } catch (ignore: UnboundException) {
                if (!klassobject.isInstance(`object`)) {
                    throw GenyrisTypeMismatchException(
                        "validator error: object " + `object` +
                                " is not tagged with " + klassobject.getClassName()
                    )
                }
            }
            if (validator != null) {
                val args: Array<Exp?>? = arrayOfNulls<Exp>(1)
                args!![0] = `object`
                val result = validator.applyFunction(klassobject, args)
                if (result === NIL) {
                    throw GenyrisException(
                        ("class " + klassobject.getClassName()
                                + " validator error for object " + `object`)
                    )
                }
            }
        }

        private fun VALIDATE(environment: Environment): SimpleSymbol? {
            return environment.getSymbolTable().VALIDATE()
        }
    }
}
